### PR TITLE
SF: Change SF_GetSweepForFormula to return also data from non-displayed sweeps

### PR DESCRIPTION
**Pull request by MichaelHuth**
Added select operation to select data from sweeps/channels. To be used with data, tp, epochs and labnotebook.

select can be used also as optional argument in the mentioned operations. With select it can be chosen to retrieve data from all or displayed sweeps.

Fixed channels operation to work correctly when specifying `channels()`.

Updated Tests, Docu.

close https://github.com/AllenInstitute/MIES/issues/1135
close https://github.com/AllenInstitute/MIES/issues/1208

### CODE REVIEW

File: Packages/MIES/MIES_SweepFormula.ipf [hunk 1]
```diff
@@ -1054,139 +1054,95 @@ static Function SF_GetDAChannel(string graph, variable sweep, variable channelTy
...
-		return $""
+	if(DimSize(range, COLS) > 0)
+		SF_ASSERT(DimSize(range, COLS) == numSweeps, "Multidimensional range column size must equal number of sweeps.")
+		SF_ASSERT(DimSize(range, LAYERS) == numChannels, "Multidimensional range layers size must equal number of channels.")
 	endif
 
-	SortColumns/A/DIML/KNDX={FindDimLabel(traces, COLS, "channelType"), FindDimLabel(traces, COLS, "channelNumber"), FindDimLabel(traces, COLS, "sweepNumber")} sortWaves=traces
+	SF_ASSERT(BSP_HasBoundDevice(graph), "No device bound.")
+	device = BSP_GetDevice(graph)
+	DFREF deviceDFR = GetDeviceDataPath(device)
```
Comment (t-b) [inline Packages/MIES/MIES_SweepFormula.ipf:1082]: I bet this fail with a sweepbrowser started from the analysis browser. SB_UpdateSweepPlot has an example how to get a sweep datafolder from a sweep browser map index.

Comment (t-b) [reply Packages/MIES/MIES_SweepFormula.ipf:1082]: This whole thing gets easy if we limit ourselves to unique sweep numbers in the sweep browser map.

**Comment by t-b** [comment]
CI: TestLabNotebook  fails

File: Packages/MIES/MIES_SweepFormula.ipf (hunk 1 above)
Comment (MichaelHuth) [reply Packages/MIES/MIES_SweepFormula.ipf]: I hope the latest changes address this.

**Comment by MichaelHuth** [comment]
@t-b Ready for a code review. 

[ ] Adapt docu

File: Packages/MIES/MIES_SweepFormula.ipf (hunk 1 above)
Comment (t-b) [reply Packages/MIES/MIES_SweepFormula.ipf]: Jeep, I'd say yes!

**Comment by t-b** [comment]
@MichaelHuth Nice! Works as expected.

- [ ] Tests

I generated some coverage info via `runwithOpts(testcase = "TestDataOperation", traceWinList = "MIES_SweepFormula.ipf", instru = 1)` and would think that the following code could benefit from some tests.

```
00000007|00000000|00000007|    if(DimSize(range, COLS) > 0)

________|________|________|        SF_ASSERT(DimSize(range, COLS) == numSweeps, "Multidimensional range column size must equal number of sweeps.")

________|________|________|        SF_ASSERT(DimSize(range, LAYERS) == numChannels, "Multidimensional range layers size must equal number of channels.")

00000007|________|________|    endif
```

```
00000013|00000000|00000013|            if(!WaveExists(sweep))

________|________|________|                continue

00000013|________|________|            endif
```

```
00000013|00000000|00000013|            if(DimSize(range, COLS) > 0)

________|________|________|                rangeStart = range[0][i][j]

________|________|________|                rangeEnd = range[1][i][j]

00000013|________|________|            else

00000013|________|________|                rangeStart = range[0]

00000013|________|________|                rangeEnd = range[1]

00000013|________|________|            endif
```

```
00000011|00000000|00000011|            if(!WaveExists(sweep))

________|________|________|                continue

00000011|________|________|            endif
```

- [ ] The documentation (rst) should mention that we take the first sweep with that number.

**Comment by MichaelHuth** [comment]
> * Clarify semantics change for `sweeps()` it now does not ignore hidden channels anymore.

I am not sure what you mean. The sweeps operation did not change. It already returned all OVS sweeps with "all" and all displayed with "displayed".
It would be possible to change the default for "sweeps()" from "dispalyed" to "all".

**Comment by t-b** [comment]
Me and @MichaelHuth chatted about `GetSweepForFormula`.

And we see the following options:
- [ ] Add optional parameter to `data` and friends to allow the user choosing between `respect ignored channels/HS in GUI` (old default) and `ignore GUI settings` (what the PR currently does)
- [ ] We could also try figuring out the above choice via an heuristic. We could use `respect` for all displayed sweeps and `ignored` for all non-displayed

**Comment by t-b** [comment]
I'ved pushed a change with the `data_displayed` operation. This now works as `data` in main. Having both available allows @timjarsky and us to figure out way to get the nice parts of both worlds.

**Comment by timjarsky** [comment]
@t-b with data_displayed and turning off a HS or HS removal:
```  BUG_TS: Encountered pending RTE: 1321, a wave read;Index out of range for wave "_free_".```

image
After moving through the error it appears to display correctly.

**Comment by timjarsky** [comment]
@t-b thanks for the fixes. Both data formulas are working nicely;

image

**Comment by t-b** [comment]
To continue this PR:

The code added a new formula `data_displayed` which works like the old `data`. And `data` now does not look at the displayed data anymore but only on the present data.

We could now just document the current state and merge but:
- The approach seems quite hacky
- It does not work with other functions which accept sweeps/channels pairs like `epochs`/`labnotebook`/`tp`

How about instead we solve it a bit differently:
- Add a new formula `selection(array channels, array sweeps, [string mode])` were `mode` accepts `displayed` or `all` and defaults to `displayed`.
- This formula would return a wave with the selected channels and sweeps
- Let `data`/`epochs`/`labnotebook`/`tp` accept a `selection` formula instead of `array channels, array sweeps`

This would have the advantage that we now have a concept of selecting existing sweeps and channels either from the displayed sweeps or all sweeps. And allow all potential consumers like `epochs` etc. to reuse that approach.

**Comment by MichaelHuth** [comment]
List:

* [ ] Introduce select operation
* [ ] Change data to use select
* [ ] Remove data_displayed
* [ ] Change tp to use select
* [ ] Change labnotebook to use select
* [ ] Change epochs to use select
* [ ] Tests
* [ ] Documentation

**Comment by t-b** [comment]
Linting failed with:
```
The duplicated newlines check failed and found the following occurences:
--
07-Apr-2022 14:05:41 | Packages/doc/SweepFormula.rst:584:`{{0, 0, 0, 0, 1, 1, 1, 1}, {0, 0, 1, 1, 0, 0, 1, 1}, {0, 1, 6, 7, 0, 1, 6, 7}}`.
07-Apr-2022 14:05:41 | Packages/doc/SweepFormula.rst:585:
07-Apr-2022 14:05:41 | Packages/doc/SweepFormula.rst:586:
07-Apr-2022 14:05:41 | Packages/doc/SweepFormula.rst:591:   select(channels(AD2, DA5, AD0, DA6), [0, 1, 3, 7], all)
07-Apr-2022 14:05:41 | Packages/doc/SweepFormula.rst:592:
07-Apr-2022 14:05:41 | Packages/doc/SweepFormula.rst:593:
```

you can run that yourself with `tools/check-code.sh`.

**Comment by t-b** [comment]
A couple of tests fail:

```
 Entering test case "PatchSeqSealEvaluation#PS_SE6:NI"
  Entering reentry "PatchSeqSealEvaluation#PS_SE6_REENTRY"
  Assumption that the first wave (wv1) exists: is false. Assertion "CHECK_EQUAL_TEXTWAVES(entries[%resultsSweep], {"0;", "1;", "2;"}, mode = WAVE_DATA)" failed in line 729, procedure "UTF_PatchSeqSealEvaluation.ipf"
  Assumption that the wave's main type is 2: is false. Assertion "CHECK_WAVE(entries[%resultsResistanceA], TEXT_WAVE)" failed in line 730, procedure "UTF_PatchSeqSealEvaluation.ipf"
  Assumption that the wave's main type is 2: is false. Assertion "CHECK_WAVE(entries[%resultsResistanceB], TEXT_WAVE)" failed in line 731, procedure "UTF_PatchSeqSealEvaluation.ipf"
  Trying to resurrect missing sweeps from device Dev1
  Sweep 0 is missing its sweep wave
  Sweep 0 is missing its config wave
  Sweep 1 is missing its sweep wave
  Sweep 1 is missing its config wave
  Sweep 2 is missing its sweep wave
  Sweep 2 is missing its config wave
  Trying to recontruct sweep 0
  Reconstructed successfully.
  Trying to recontruct sweep 1
  Reconstructed successfully.
  Trying to recontruct sweep 2
  Reconstructed successfully.
  Leaving test case "PatchSeqSealEvaluation#PS_SE6:NI"
```

[JU_2022_04_07_170329_2022-04-07_17-54-42.zip](https://github.com/AllenInstitute/MIES/files/8451530/JU_2022_04_07_170329_2022-04-07_17-54-42.zip)

This looks due to the `store` calls needing adaptation.

**Comment by MichaelHuth** [comment]
Add SF_SetFormula -> sets formula in DB/SB
Add SF_ExecuteFormula -> executed given formula string and returns wave

**Comment by MichaelHuth** [comment]
@t-b I changed the order in the test in the last commit. I think it is ok, but please verify. You probably know better if we have something that depends on the order of the entries in the "Sweep Formula active channels" labnotebook entry.

**Comment by t-b** [comment]
@MichaelHuth Thx. Will have a look.

**Comment by t-b** [comment]
First test:
- It is hideously slow. You are calling in SF_GetActiveChannelNumbers the function GetLastSettingChannel in the most inner loop with three loops. This is unnecessary as you don't need to call GetLastSettingChannel for every channel number.
- My commit but still GetLastSettingChannel  is way too slow by itself already due to FindIndizes. A plain GetRowIndex does the job as well.

**Comment by timjarsky** [comment]
@MichaelHuth @t-b with long sweeps (60s) it is slow. We can leave this as a known issue if it is challenging to resolve.

```
Total time: 11.3524, Time in Function code: 2.87061 (25.3%)
Top function percentages:
Function MIES_SweepFormula.ipf SF_GetSweepsForFormula: 82%

Annotated Top Functions:

*******************************************************************************************
Function: MIES_SweepFormula.ipf SF_GetSweepsForFormula; Percent total 82%
*******************************************************************************************
[00]*         	|static Function/WAVE SF_GetSweepsForFormula(string graph, WAVE range, WAVE selectData)
[00]          	|
[00]          	|	variable i, j, rangeStart, rangeEnd, pOffset, delta, numRows, DAChannel, numSweeps, sweepNo
[00]          	|	variable chanNr, chanType, numChannels, indexOffset, iStart, iLength, cIndex, isSweepBrowser
[00]          	|	variable dimPos, numChannelTypes, numChannelNumbers, index
[00]          	|	string dimLabel, device, dataFolder
[00]          	|
[00]          	|	ASSERT(WindowExists(graph), "graph window does not exist")
[00]          	|	if(IsTextWave(range))
[00]          	|		SF_ASSERT(DimSize(range, ROWS) == 1, "A epoch range must be a single string with the epoch name.")
[00]          	|		WAVE/T wEpochName = range
[00]          	|	else
[00]          	|		SF_ASSERT(DimSize(range, ROWS) == 2, "A numerical range is must have two rows for range start and end.")
[00]          	|	endif
[00]          	|	if(SF_IsDefaultEmptyWave(selectData))
[00]          	|		return $""
[00]          	|	endif
[00]          	|	SF_ASSERT(DimSize(selectData, COLS) == 3, "Select data must have 3 columns.")
[00]          	|
[00]*         	|	WAVE/Z sweeps
[00]          	|	WAVE/Z activeChannels
[00]          	|	[sweeps, activeChannels] = SF_ReCreateOldSweepsChannelLayout(selectData)
[00]          	|
[00]          	|	numSweeps = DimSize(sweeps, ROWS)
[00]          	|	numChannels = DimSize(activeChannels, ROWS)
[00]          	|
[00]          	|	if(DimSize(range, COLS) > 0)
[00]          	|		SF_ASSERT(DimSize(range, COLS) == numSweeps, "Multidimensional range column size must equal number of sweeps.")
[00]          	|		SF_ASSERT(DimSize(range, LAYERS) == numChannels, "Multidimensional range layers size must equal number of channels.")
[00]          	|	endif
[00]          	|
[00]          	|	isSweepBrowser = BSP_IsSweepBrowser(graph)
[00]          	|
[00]          	|	if(isSweepBrowser)
[00]*         	|		DFREF sweepBrowserDFR = SB_GetSweepBrowserFolder(graph)
[00]          	|		WAVE/T sweepMap = GetSweepBrowserMap(sweepBrowserDFR)
[00]          	|	else
[00]          	|		SF_ASSERT(BSP_HasBoundDevice(graph), "No device bound.")
[00]          	|		device = BSP_GetDevice(graph)
[00]          	|		DFREF deviceDFR = GetDeviceDataPath(device)
[00]          	|	endif
[00]          	|
[00]*         	|	Make/FREE/WAVE/N=(numSweeps, numChannels) sweepRefs
[00]*         	|	Make/FREE/U/I/N=(numSweeps, numChannels) indexStart, indexEnd
[00]          	|
[00]          	|	for(i = 0; i < numSweeps; i += 1)
[00]          	|
[00]          	|		sweepNo = sweeps[i]
[00]          	|
[00]          	|		if(isSweepBrowser)
[00]*         	|			cIndex = FindDimLabel(sweepMap, COLS, "Sweep")
[00]*         	|			FindValue/RMD=[][cIndex]/TEXT=num2istr(sweepNo)/TXOP=4 sweepMap
[00]          	|			if(V_value == -1)
[00]          	|				continue
[00]          	|			endif
[00]*         	|			dataFolder = sweepMap[V_row][%DataFolder]
[00]*         	|			device     = sweepMap[V_row][%Device]
[00]*         	|			DFREF deviceDFR  = GetAnalysisSweepPath(dataFolder, device)
[00]          	|		endif
[00]*         	|		DFREF sweepDFR = GetSingleSweepFolder(deviceDFR, sweepNo)
[00]          	|
[00]          	|		for(j = 0; j < numChannels; j += 1)
[00]          	|
[00]          	|			chanNr = activeChannels[j][%channelNumber]
[00]*         	|			chanType = activeChannels[j][%channelType]
[00]*         	|			WAVE/Z sweep = GetDAQDataSingleColumnWave(sweepDFR, chanType, chanNr)
[00]          	|			if(!WaveExists(sweep))
[00]          	|				continue
[00]          	|			endif
[00]          	|
[00]          	|			if(delta == 0)
[00]          	|				delta = DimDelta(sweep, ROWS)
[00]          	|			else
[00]          	|				SF_ASSERT(delta == DimDelta(sweep, ROWS), "Sweeps have not the same delta.")
[00]          	|			endif
[00]          	|
[00]          	|			if(WaveExists(wEpochName))
[00]          	|				DAChannel = SF_GetDAChannel(graph, sweepNo, chanType, chanNr)
[00]          	|				WAVE range = SF_GetRangeFromEpoch(graph, wEpochName[0], sweepNo, DAChannel)
[00]          	|			endif
[00]          	|
[00]          	|			if(DimSize(range, COLS) > 0)
[00]          	|				rangeStart = range[0][i][j]
[00]          	|				rangeEnd = range[1][i][j]
[00]          	|			else
[00]          	|				rangeStart = range[0]
[00]          	|				rangeEnd = range[1]
[00]          	|			endif
[00]*         	|			SF_ASSERT(!IsNaN(rangeStart) && !IsNaN(rangeEnd), "Specified range not valid.")
[00]          	|			SF_ASSERT(rangeStart == -inf || (IsFinite(rangeStart) && rangeStart >= leftx(sweep) && rangeStart < rightx(sweep)), "Specified starting range not inside sweep " + num2istr(sweepNo) + ".")
[00]*         	|			SF_ASSERT(rangeEnd == inf || (IsFinite(rangeEnd) && rangeEnd >= leftx(sweep) && rangeEnd < rightx(sweep)), "Specified ending range not inside sweep " + num2istr(sweepNo) + ".")
[00]          	|
[00]*         	|			indexStart[i][j] = ScaleToIndexWrapper(sweep, rangeStart, ROWS)
[00]          	|			indexEnd[i][j] = ScaleToIndexWrapper(sweep, rangeEnd, ROWS)
[00]*         	|			sweepRefs[i][j] = sweep
[00]          	|		endfor
[00]          	|	endfor
[00]          	|
[00]          	|	indexOffset = WaveMin(indexStart)
[00]          	|	numRows = WaveMax(indexEnd) + 1 - indexOffset
[00]          	|
[00]          	|	// combine sweeps to data wave
[66]*******   	|	Make/FREE/N=(numRows, numSweeps, numChannels) sweepData = NaN
[00]          	|	SetScale/P x, indexOffset * delta, delta, sweepData
[00]          	|	for(i = 0; i < numSweeps; i += 1)
[02]*         	|		for(j = 0; j < numChannels; j += 1)
[00]          	|
[00]*         	|			WAVE/Z sweep = sweepRefs[i][j]
[00]          	|			if(!WaveExists(sweep))
[00]          	|				continue
[00]          	|			endif
[00]          	|
[00]*         	|			iStart = indexStart[i][j]
[00]          	|			iLength = indexEnd[i][j] - iStart
[00]          	|			pOffset = iStart - indexOffset
[13]*         	|			MultiThread sweepData[pOffset, pOffset + iLength][i][j] = sweep[iStart + p - pOffset]
[00]          	|
[00]          	|			if(i == 0)
[00]          	|				dimLabel = StringFromList(activeChannels[j][%channelType], XOP_CHANNEL_NAMES) + num2istr(activeChannels[j][%channelNumber])
[00]          	|				SetDimLabel LAYERS, j, $dimLabel, sweepData
[00]          	|			endif
[00]          	|		endfor
[00]          	|
[00]*         	|		sprintf dimLabel, "sweep%d", sweeps[i]
[00]*         	|		SetDimLabel COLS, i, $dimLabel, sweepData
[00]          	|	endfor
[00]          	|
[00]          	|	return sweepData
[00]          	|End
```

**Comment by MichaelHuth** [comment]
I have made a small change that should speed up the initialization with NaN for bigger sweeps.

**Comment by t-b** [comment]
Review:

7baccefc9 (SF: Change SF_GetSweepForFormula to return also data from non-displayed sweeps, 2022-01-19)

Jeep.

4d1af70e1 (Adapt CreateFakeSweepData to PrepareLBN_Ignore changes, 2022-01-21)

- [ ] Prototype functions should assert out in production code and `FAIL()` in test code. We don't want to call them ever.

6a25cb68d (SF: Add Tests for data operation, 2022-01-21)
+1

02714cf28 (SF: Add select operation to select sweeps for data, adapted data operation, 2022-02-04)

- [ ] SF_OperationSelect check that the third parameter is text, but it does not check if it is either `all` or `displayed`.
- [ ] Create a standard wave getter for selectData with documentation about its structure. Right now the code to create it is duplicated.
- [ ] SF_ReCreateOldSweepsChannelLayout and SF_GetReducedColumn need comments what they do. Please also prefer `GetUniqueEntries` over FindDuplicates as that handles the details already.
- [ ] SF_IsDefaultEmptyWave adapt to

```
if(DimSize(w, ROWS) == 1 && !DimSize(w, COLS) && IsNumericWave(w))
```

and add comment.

735fba1f7 (SF: Adapt tp operation to use select instead of sweeps and channels, 2022-04-05)

Jeep.

1345a8ca8 (SF: Change labnotebook operation to use select instead of sweeps, channels, 2022-04-05)

+1

9f48adf3b (SF: Changes epochs operation to use select instead of channels, sweeps, 2022-04-05)

+1 
7dd83a27e (SF tests: Adapt Epochs Tests for using select operation, 2022-04-05)

+1
847a6c5c0 (SF Tests: Adapt labnotebook operation tests using select(...), 2022-04-05)

+1

a344a34f4 (SF tests: Adapt test AvoidAssertingOutWithNoSweeps using select operation, 2022-04-05)

+1

3aee1299a (SF Tests: Adapt data operation tests using select, 2022-04-05)

+1

3f1ec9e27 (SF: Adapts tests for tp operation for select introduction, 2022-04-07)

+1

d03b3bf6f (DB: Adapt default SF notebook template with data using select, 2022-04-07)

Jeep.
ec7219c67 (SF: Add tests for operation select(...), 2022-04-07)

- [ ] I've now ran the tests with tracing `runwithOpts(testsuite= "UTF_SweepFormula.ipf", instru=1, traceWinList = "MIES_SweepFormula.ipf")` and found a couple of places where we could use some more tests.

```
1223|00000009|00000000|00000009|    if(SF_IsDefaultEmptyWave(selectData))

1224|________|________|________|        return $""

1225|00000009|________|________|    endif
```

```
1317|00000017|00000000|00000017|            if(!WaveExists(sweep))

1318|________|________|________|                continue

1319|00000017|________|________|            endif
```

```
1275|00000019|00000000|00000019|            if(!WaveExists(sweep))

1276|________|________|________|                continue

1277|00000019|________|________|            endif
```

eac53af12 (Docu: Update SF docu for select and changes of data, labnotebook, epochs, tp, 2022-01-21)

```
+Through the `select` function it can be chosen of data for sweeps is returned for
+currently displayed sweeps or for all existing sweeps.
```

- [ ] I can't parse that. Feel free to ignore if it is fixed in a later commit.

26ad02355 (SF: Update docu in SF notebook for the new select function., 2022-04-07)

Nice!

fb6180fd4 (SF: Change sweeps operation to use zero parameters and return all sweeps, 2022-04-08)

good.

07f5ffa43 (Updated SF example in code docu comment in MIES_AnalysisFunctions.ipf, 2022-04-08)

Nice find.

ff6ae88da (SF: Adapt store operation to replace use of sweeps(displayed), 2022-04-08)

- [ ] Prefer `GetUniqueEntries`
- [ ] SF_CreateResultsWaveWithCode

26275a18c (Updated comment for todo in AnalysisFunctions PSQ, 2022-04-08)

Jeep.

41317c24e (SF: Adapt tests for sweeps() operation working on all sweeps now, 2022-04-08)

- [ ] Adapting OVS_GetSelectedSweeps should be in its own commit. This is a pure cleanup.
- [ ] Prefer `GetUniqueEntries`

SF_CreateResultsWaveWithCode:
- [ ] `channels(AD)` to `channels()`
- [ ] I also think that we don't need the second channels() call as we now have everything with the select call or?

7cef53c2a (SF: bugfix channels operation, now properly returning NaNs, 2022-04-08)

Nice!

b90214f89 (SF: Adapt tests for channels() operation for previous fix, 2022-04-09)

Jeep.

Look into tracing again.

714425942 (SF: Docu adapted documentation for the fixed channels operation, 2022-04-09)

Current state does not compile.

0afee08d7 (SF: Add Utility functions to set a formula in SF NB and to execute a formula, 2022-04-09)
46014df48 (DataBrowser: Use SF_SetFormula in DB and update SF notebook initial string, 2022-04-09)
2acfad296 (SF: use SF_SetFormula and SF_ExecuteFormula in other MIES code, 2022-04-09)
cfecaca77 (SF: Allow select operation to be called with no arguments, 2022-04-09)

All look good.

fad19026b (SF: Hardened SF_IsDefaultEmptyWave utility function., 2022-04-09)

Fold into commit introducing the code.

fe64f4fd6 (SF: Add test for select operation without arguments select(), 2022-04-09)

Good.

89e230d11 (SF: Add Tests for data without optional arguments., 2022-04-09)

Good.

5f6d75628 (SF: Change in tp operation that select is optional, 2022-04-10)

Good.

62eba0bcf (SF: Add tests for optional select in tp operation, 2022-04-10)

Good.

677d307d7 (SF: Change operation epochs to allow select argument to be optional, 2022-04-10)

Good.

3d4633709 (SF: Add Tests for operation epochs optional select argument, 2022-04-10)


- [ ] 

```diff
-        WAVE data = SF_FormulaExecutor(DirectToFormulaParser(str), graph = win); AbortOnRTE
+        WAVE data = SF_FormulaExecutor(DirectToFormulaParser(str), graph = win)
```

If that change is on purpose, it should be in its own commit and justified.

4aa668f67 (SF: Changes labnotebook operation for optional select argument, 2022-04-10)

Good.

5a6f064d2 (SF: Add Test for optional select argument of labnotebook operation, 2022-04-11)

Good.

21fdd2558 (SF: Adapted SweepFormula documentation for optional select argument, 2022-04-11)
eb4f75244 (SF: Update SweepFormula help notebook with optional select argument description, 2022-04-11)

All good.

b852f00be (SF: Adapted TestSweepFormulaCodeResults test for changed order active channels, 2022-04-11)

Changing the order is okay.

1f776003f (Optimize speed of GetLastSettingInternal, 2022-04-12)
8312480d9 (SF: Optimize SF_GetActiveChannelNumbers, no GetLastSettingChannel calls, 2022-04-12)
646d9c45b (SF: Speedup memory fill for larger sweep data in SF_GetSweepFormula, 2022-04-13)

Jeep.

So I think functionality wise this is good to go. But I still like to see some performance improvements.

For example take an arbitrary sweep with a 20ms intersted tp, use `tp(0)` as formula and do function profiling.

I then get:
```
Total time: 2.53212, Time in Function code: 0.130048 (5.14%)
Top function percentages:
Function MIES_MiesUtilities.ipf FindIndizes: 22%
Function MIES_SweepFormula.ipf SF_FormulaPlotter: 13%
Function MIES_MiesUtilities.ipf GetLastSetting: 11%
Function MIES_Utilities.ipf ZapNaNs: 7%
Function MIES_SweepFormula.ipf SF_GetSweepsForFormula: 7%
Function MIES_Utilities.ipf createDFWithAllParents: 3%
Function MIES_Utilities.ipf GetNumberFromWaveNote: 3%
Function MIES_WaveDataFolderGetters.ipf UpgradeDataFolderLocation: 2%
Function MIES_Cache.ipf CA_GetCacheIndex: 2%
Function MIES_WaveDataFolderGetters.ipf GetCacheKeyWave: 2%
Function MIES_Cache.ipf CA_TryFetchingEntryFromCache: 2%
Function MIES_SweepFormula.ipf SF_GetActiveChannelNumbers: 2%
Function MIES_MiesUtilities.ipf ParseDeviceString: 2%
Function MIES_WaveDataFolderGetters.ipf GetDevSpecLabNBFolderAsString: 1%
Function MIES_Utilities.ipf WindowExists: 1%
Function MIES_BrowserSettingsPanel.ipf BSP_GetDevice: 1%
```

This is with 100 sweeps in total and 2 channels.

SF_GetSelectDataForDisplayed calls FindIndizes in three locations. I think it should be faster to iterate over `traces` as that is always less or equal to sweeps. The SortColumns call can then also be dropped.

`SF_OperationSelect` calls SF_GetActiveChannelNumbers first, so this will always be with all available sweeps. But I'd say that SF_GetSelectDataForDisplayed and SF_GetSelectDataForAll can both ignore non-existant channels by themselves. The first by iterating over traces and the second already does that. And so by removing SF_GetActiveChannelNumbers we should also be able to get rid of the 11% from GetLastSetting.

I'm happy to do the performance improvements in a follow up PR.

**Comment by t-b** [comment]
@MichaelHuth I think this PR also solves #1135 or?

**Comment by t-b** [comment]
@timjarsky I had only some speed and code nits. Is it functionality wise what you expect?

**Comment by timjarsky** [comment]
@t-b I think so? I played with it quite a bit yesterday and generally only noted unrelated issues. In one case, I was less sure; it seemed like using the shortcuts for epoch info wasn't working using a test file that was new enough to have short names. I can demo tomorrow, if you are available?

**Comment by MichaelHuth** [comment]
So I overhauled the code and it should be faster now.

**Comment by t-b** [comment]
@timjarsky We had one more idea to make it faster. This is now worked on.

**Comment by timjarsky** [comment]
> @timjarsky We had one more idea to make it faster. This is now worked on.

@t-b I will test today.

**Comment by MichaelHuth** [comment]
Ok should be even faster now. Could you check once if for the most common use case the optimization boil down at L1326 helps?
To remove comment it out until the else.

**Comment by t-b** [comment]
We always get the following:

grafik

**Comment by MichaelHuth** [comment]
@t-b Should be fixed now.

**Comment by t-b** [comment]
Thx.

Yes now are in a good range:
```
Total time: 2.44164, Time in Function code: 0.0457284 (1.87%)
Top function percentages:
Function MIES_SweepFormula.ipf SF_FormulaPlotter: 42%
Function MIES_SweepFormula.ipf SF_GetSweepsForFormula: 19%
Function MIES_MiesUtilities.ipf GetLastSetting: 4%
```

with `tp(0)` with 100 sweeps and one displayed.

**Review by t-b — APPROVED** [review]

**Comment by timjarsky** [comment]
@t-b long sweeps are much quicker (>2X). Thanks!

```
Total time: 5.57715, Time in Function code: 1.18348 (21.2%)
Top function percentages:
Function MIES_SweepFormula.ipf SF_GetSweepsForFormula: 83%
```

**Review by timjarsky — APPROVED** [review]

**Comment by t-b** [comment]
I've fixed conflicts, will merge once CI passes.

**Pull request merged**